### ISSUE TITLE
vector histograms not being created with new THaFormula, THaCuts, THaVform

### ISSUE THREAD

**Issue by rwmichaels** [JeffersonLab/analyzer#62]
When I merged to the main branch of the analyzer, i.e the version supported by Ole, I found that the feature of "vector histograms" did not work.  This is a feature whereby a histogram for a vector of variables, for example for the shower counter, would expand to a vector of histograms like Rsha_p0 ... Rsha_p79 (there are 80 of those blocks).  It's more convenient than defining 80 histograms in the output.def file.  Currently you get no histograms at all, and the line in output.def is reported as an error.  What I've observed so far is that if I copy over the tagged-release-156 version of the following classes, the feature is restored.  I don't understand WHY yet, but I'm investigating. Those classes are THaFormula, THaCut, and THaVform.

**Comment by hansenjo** [comment]
Can you give and example line from output.def that shows this error?

Ole

**Comment by rwmichaels** [comment]
Example of problematic line in output.def

TH1F Ls2rt 'Left arm S2 R-PMT TDCs' L.s2.rt 750 1001 2500 L.s2.rt>0.

Normally this produces a vector of histograms like Ls2rt0, Ls2rt1,Ls2rt2,  Ls2rt4
and you can draw them like

Ls2rt4->Draw()

Now, however, you get an error message when running Podd, and no such histograms.

I should mention there is a dirty-trick assumption in the interpretation of that line

If you did 
TH1F Ls2rt 'Left arm S2 R-PMT TDCs' L.s2.rt 750 1001 2500

or if you had a "scalar" cut like  then JustOneTrackL, then all the contents of the array Ls2rt get put into one histogram called Ls2rt.  

However with the "vector" cut at the end, like L.s2.rt>0., then you get this expansion into a vector
of histograms, which most experimenters find convenient.

I recall that historically this only worked for "Fixed Size" arrays.  If the arrays were variable sized, the code punts, issues an error message, and no histograms.  I think that is what is happening now.  I am close to finding a solution if you want to let me continue trying.

**Comment by hansenjo** [comment]
Thanks for the example. That'll help.

Quick comment: at least in the 1.6 version, a THaCut is always a scalar.
Not sure if that matters, maybe you implemented your own "vector cuts".

Variable-size arrays are now supported in THaFormula, although it may
not be easy to make them work with vector histograms. How many
histograms to create is not known at initialization time. Variable-size
arrays report size zero at that time.

I'd be very happy to look into fixing this (I broke it!), but I'm
currently swamped with preparations for the next review. I could start
in two weeks.

BTW, there is also a request to make the cuts defined in the output
package available in the gHaCuts list, so their statistics can be
accessed at the end of a run, for example. Steve Wood has this nice
configurable "report" feature that prints all kinds of useful
information about a run. These cuts would have to be put in a new cut
block called, say, "Output".

**Comment by rwmichaels** [comment]
This issue has been resolved.  The solution is shown by the diff below for THaVform.C.  I've made this change in my branch mergeUpstream.  This change restores the original code.

As Ole explained, variables like L.s2.lt are technically "variable sized" but if they were initialized with nonzero length they are fixed for the duration of the run.  

in THaVform::DefinedGlobalVariable( TString& name )

>     } else if( gvar->IsVarArray() ) {
>       stat = kVAType;
739,742d739
<     if (gvar->GetLen() == 0) {
<       stat = kVAType;
<     }
